### DISCUSSION TITLE
Support for Heat or BTU Meter

### DISCUSSION THREAD

**Discussion by nagyrobi** [arendst/Tasmota#11128] [Ideas]
Is there any heat meter (n.b. not temperature, but heat, as heat quantity meter or flux calorimeter) device supported by Tasmota in any configuration? Like the Smart Meter Interface?

If not, in theory this could be implemented in software using discrete components:

A heat meter consists of (from [wiki](https://en.m.wikipedia.org/wiki/Heat_meter)):
  -  a fluid flow meter - typically a turbine-type flow meter, or alternatively an ultrasonic flow meter;
  -  a means of measuring the temperature between the outflow and the inflow - usually a pair of temperature sensors;
  -  a means of integrating the two measurements over a period of time - typically half an hour - and accumulating the total heat transfer in a given period.

Any ideas?

Tips: https://www.letscontrolit.com/forum/viewtopic.php?t=2790

**Comment by sfromis** [comment]
With the 3 sensors, the calculations should not be too difficult to implement, either in rules, or if more complex, maybe in backend.

**Reply by nagyrobi** [reply]
Moreover, in certain cases you don't even need a flow meter. When you use only 1 pump, and you want to measure the heat energy transferred through it, it's enough to look at the time durations while the pump was switched on and off. The pump has a spec sheet from factory which states the exact flow it produces so only having a Tasmota device with 2 temp sensors and a switch commanding the pump, could do the job. The flow produced by the pump should be entered as a constant value.

The question is how to implement the math.

**Reply by javier6d2** [reply]
> The question is how to implement the math.

Scripting something like:
```

>D
var1=0
var2=0
var3=120 ;yours constant flow, L/h units
var4=0 ;heat energy kW
>T
; get sensor temp values,
var1=DS18B20-1#Temperature
var2=DS18B20-2#Temperature
>S
if mqtts>0
and upsecs%tper==0
then
var4=4.18*var3*(var2-var1)/3600
=>Publish tele/%topic%/…something more
endif
#
```

**Reply by nagyrobi** [reply]
Thanks. Wonder how to take switch into consideration.
So only calculate while switch was on (pump was running).

**Reply by javier6d2** [reply]
> Wonder how to take switch into consideration.

May be you can try compiling  with USE_SML_SCRIPT_CMD, using Conditional Statements and >S (Executed every second) Script Section. Not sure, but something like:

```
var5= %sml[5]% ;Solar pump R1 

If  var5 = 1 
then
var4+=4.18*var3*(var2-var1)/3600
```

**Reply by nagyrobi** [reply]
Not good because the math should contain the fraction of time passed since the the pump turned on.

**Reply by javier6d2** [reply]
Yes, my output is just Power and you want Energy or Heat.

**Reply by nagyrobi** [reply]
Yep.

**Reply by javier6d2** [reply]
You can try:

```
>D
pump=0
T1=0
T2=0
flow=120; yours constant flow, L/h units
power=0
count=0
heat=0
heat_d=0
hour=3600
day=86400
>S
T1=DS18B20-1#Temperature
T2=DS18B20-2#Temperature
pump= get your pump status
if pump==1
then
power+=4.18*flow*(T2-T1)/3600
count+=1
else
count+=1
endif
if upsecs%hour==0
then
heat+=power/count
print energy delivered per hour %heat%kwh
heat_d+=heat
hour==0
heat=0
count=0
power=0
endif
if upsecs%day==0
then	
print energy delivered per day %heat_d%kwh %heat%
day==0
heat_d=0
endif
#

```

**Reply by nagyrobi** [reply]
Cool, I'll check.

**Reply by malinduta** [reply]
Hello!
How it is working, @nagyrobi?
I would like to implement it too.

**Reply by nagyrobi** [reply]
Moved to ESPHome...

**Reply by malinduta** [reply]
How about the math? It is working as expected?
Never used ESPHome.

**Reply by nagyrobi** [reply]
Can't tell as don't use it...
